From 6ce45a1106f5eab236c894bab23e8160e22c40a4 Mon Sep 17 00:00:00 2001
From: Guy Nisan <gnisan@habana.ai>
Date: Mon, 01 Nov 2021 10:17:30 +0200
Subject: [PATCH] [SW-42057] gaudi2: print hbm info in fuse

Print out the HBM info stored in fuse. The printed fields are "HBM
redudancy" "SLT" and "SLT done".

Change-Id: Icc5ccd0b84f29fd26fd0783de8a8e9cc0dd97c5e
Signed-off-by: Guy Nisan <gnisan@habana.ai>
---

diff --git a/arch/gaudi2/common/asic.c b/arch/gaudi2/common/asic.c
index 7ca0343..eccb5af 100644
--- a/arch/gaudi2/common/asic.c
+++ b/arch/gaudi2/common/asic.c
@@ -18,6 +18,7 @@
 #include <uart.h>
 #include <pci.h>
 #include <sha.h>
+#include <efuse.h>
 
 #define HBM_BAR_PF_NUM (1 << 4)
 
@@ -204,3 +205,46 @@
 {
 	sha384_csum(input, ilen, output, 0);
 }
+
+#define FUSE_HBM_REDUNDANCY_WORD	18
+#define FUSE_HBM_REDUNDANCY_MASK	0x0003F000
+#define FUSE_HBM_REDUNDANCY_SHIFT	12
+
+#define FUSE_HBM_SLT_WORD		17
+#define FUSE_HBM_SLT_MASK		0x7E000000
+#define FUSE_HBM_SLT_SHIFT		25
+#define FUSE_HBM_SLT_DONE_MASK		0x80000000
+#define FUSE_HBM_SLT_DONE_SHIFT		31
+
+static void hl_fuse_print_hbm(void)
+{
+	u32 val, slt, red;
+	bool slt_done;
+	int err;
+
+	err = preboot_fuse_read(FUSE_HBM_REDUNDANCY_WORD, &val);
+	if (err)
+		goto fuse_hbm_print_err;
+
+	red = (val & FUSE_HBM_REDUNDANCY_MASK) >> FUSE_HBM_REDUNDANCY_SHIFT;
+
+	err = preboot_fuse_read(FUSE_HBM_SLT_WORD, &val);
+	if (err)
+		goto fuse_hbm_print_err;
+
+	slt = (val & FUSE_HBM_SLT_MASK) >> FUSE_HBM_SLT_SHIFT;
+	slt_done = !!(val & FUSE_HBM_SLT_DONE_MASK);
+
+	preboot_puts_hex_raw("HBM redundancy: ", red, false);
+	preboot_puts_hex_raw(" slt: ", slt, false);
+	preboot_puts_hex(" slt_done: ", slt_done, false);
+
+fuse_hbm_print_err:
+	if (err)
+		preboot_puts("HBM fuse data is not available!\n");
+}
+
+void hl_ram_init(void)
+{
+	hl_fuse_print_hbm();
+}
